src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
```

Comment:
???
Looks like we talk about different subjects...

The `Recover` may work only when we bubble exception into the `RetryTemplate`.
In our case the `onMessage()` method should do that.
But since `onMessage()` calls `MessageProducerSupport.sendMessage()` and that one consults `errorChannel` we will never have  `Recover` working. Move over - we don't have even retry working.

On the other hand what you are saying doesn't fit  error channel algorithm as well.
If `recoverer` throws exceptions it is already far away from the `MessageProducerSupport` with its `errorChannel` and the exception is bubbled to the `AbstractMessageListenerContainer` directly.

I am so sorry if I miss or just don't understand something, but I just don't feel confident how feature should work. Therefore all this messy comments 😢 

We can merge it as is and continue discussion in the separate issue if that.